ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/MiniProtocol/ChainSync/Client.hs

```diff
@@ -244,17 +252,28 @@ bracketChainSyncClient
 
     -- | Wrapper around 'LeakyBucket.execAgainstBucket' that handles the
     -- disabled bucket by running the given action with dummy handlers.
-    lopBucketConfig :: MonadThrow m => LeakyBucket.Config m
-    lopBucketConfig =
-      case csBucketConfig of
-        ChainSyncLoPBucketDisabled -> LeakyBucket.dummyConfig
-        ChainSyncLoPBucketEnabled ChainSyncLoPBucketEnabledConfig {csbcCapacity, csbcRate} ->
+    lopBucketConfig :: GsmState -> LeakyBucket.Config m
+    lopBucketConfig gsmState =
+      case (gsmState, csBucketConfig) of
+        (Syncing, ChainSyncLoPBucketEnabled ChainSyncLoPBucketEnabledConfig {csbcCapacity, csbcRate}) ->
             LeakyBucket.Config
               { capacity = fromInteger $ csbcCapacity,
                 rate = csbcRate,
                 onEmpty = throwIO EmptyBucket,
                 fillOnOverflow = True
               }
+        -- NOTE: If we decide to slow the bucket down when “almost caught-up”,
+        -- we should add a state to the GSM and corresponding configuration
+        -- fields and a bucket config here.
+        _ -> LeakyBucket.dummyConfig
+
+    -- | Follow changes in the GSM state by updating the configuration of the
+    -- LoP bucket.
+    lopBucketGsmUpdater :: GsmState -> (LeakyBucket.Config m -> m ()) -> m ()
+    lopBucketGsmUpdater prevGsmState setConfig = do
```

Comment:
Hmm, we might be able to avoid blocking on GSM state changes (although they shouldn't be too frequent) by passing in the `LeakyBucket` from outside, and also updating the bucket config in the callback invoked by the GSM logic.

Reply:
That's exactly the kind of comments I was looking for. I was wondering which approach to take. It seemed to me that passing the bucket to the GSM was the more natural one but broke the scoping of the LoP being local to the ChainSync client. I suppose we could also have something like the varCandidates but for the buckets, where each peer registers their bucket in a TVar and the GSM just updates them all at once.